diff --git a/package/base-files/files/bin/config_generate b/package/base-files/files/bin/config_generate
index ee3958e..8848aaf 100755
--- a/package/base-files/files/bin/config_generate
+++ b/package/base-files/files/bin/config_generate
@@ -100,8 +100,6 @@ generate_network() {
 		json_select ..
 	json_select ..
 
-	[ -n "$ifname" ] || return
-
 	# force bridge for multi-interface devices (and lan)
 	case "$1:$ifname" in
 		*\ * | lan:*) type="bridge" ;;
diff --git a/package/base-files/files/etc/board.d/99-default_network b/package/base-files/files/etc/board.d/99-default_network
index d70aa35..a091f3f 100755
--- a/package/base-files/files/etc/board.d/99-default_network
+++ b/package/base-files/files/etc/board.d/99-default_network
@@ -9,7 +9,8 @@ board_config_update
 
 json_is_a network object && exit 0
 
-ucidef_set_interface_lan 'eth0'
+ucidef_set_interface 'lan'
+[ -d /sys/class/net/eth0 ] && ucidef_set_interface_lan 'eth0'
 [ -d /sys/class/net/eth1 ] && ucidef_set_interface_wan 'eth1'
 
 board_config_flush
diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
index 1e522b9..2f2c464 100755
--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
+++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
@@ -444,6 +444,8 @@ ath79_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0@eth0" "3:lan:1" "4:lan:2"
 		;;
+	zsun,sd100)
+		;;
 	*)
 		ucidef_set_interfaces_lan_wan "eth0" "eth1"
 		;;
diff --git a/target/linux/ath79/generic/base-files/etc/uci-defaults/95_enable-wifi b/target/linux/ath79/generic/base-files/etc/uci-defaults/95_enable-wifi
new file mode 100644
index 0000000..853b7cf
--- /dev/null
+++ b/target/linux/ath79/generic/base-files/etc/uci-defaults/95_enable-wifi
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+[ ! -z "$(uci -q get network.lan.ifname)" ] && exit 0
+
+uci set wireless.@wifi-device[0].disabled=0
+uci commit wireless
+
+exit 0
diff --git a/target/linux/ath79/generic/base-files/etc/rc.button/BTN_0 b/target/linux/ath79/generic/base-files/etc/rc.button/BTN_0
new file mode 100755
index 0000000..ae74e1f
--- /dev/null
+++ b/target/linux/ath79/generic/base-files/etc/rc.button/BTN_0
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+case "$ACTION" in
+pressed)
+	logger -t sdcard_event "SD Card inserted"
+	echo 1 > /sys/class/gpio/sdreader_reset/value
+	echo 0 > /sys/class/gpio/sdreader_reset/value
+	;;
+released)
+	logger -t sdcard_event "SD Card removed"
+	echo 1 > /sys/class/gpio/sdreader_reset/value
+	;;
+esac
+
+return 0
diff --git a/target/linux/ath79/dts/ar9331_zsun_sd100.dts b/target/linux/ath79/dts/ar9331_zsun_sd100.dts
new file mode 100644
index 0000000..47f0ffc
--- /dev/null
+++ b/target/linux/ath79/dts/ar9331_zsun_sd100.dts
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar9331.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "Zsun SD100";
+	compatible = "zsun,sd100", "qca,ar9331";
+
+	aliases {
+		led-boot = &led_status;
+		led-failsafe = &led_status;
+		led-running = &led_status;
+		led-upgrade = &led_status;
+		serial0 = &uart;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_status: status {
+			label = "sd100:green:status";
+			gpios = <&gpio 0 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		sdcard_event {
+			label = "sdcard_event";
+			linux,code = <BTN_0>;
+			gpios = <&gpio 22 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		sdreader_reset {
+			gpio-export,name = "sdreader_reset";
+			gpio-export,output = <1>;
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+		};
+
+		sdreader_switch {
+			gpio-export,name = "sdreader_switch";
+			gpio-export,output = <0>;
+			gpios = <&gpio 21 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		spi-max-frequency = <50000000>;
+		reg = <0>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x010000>;
+				read-only;
+			};
+
+			partition@10000 {
+				label = "u-boot-env";
+				reg = <0x010000 0x010000>;
+				read-only;
+			};
+
+			partition@20000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x020000 0xbc0000>;
+			};
+
+			partition@be0000 {
+				label = "recovery";
+				reg = <0xbe0000 0x400000>;
+				read-only;
+			};
+
+			partition@fe0000 {
+				label = "nvram";
+				reg = <0xfe0000 0x010000>;
+				read-only;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+};
+
+&usb {
+	status = "okay";
+
+	compatible = "generic-ehci";
+	reg = <0x1b000000 0x1000>;
+
+	has-transaction-translator;
+	caps-offset = <0x100>;
+};
+
+&usb_phy {
+	status = "okay";
+};
+
+&wmac {
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x1002>;
+};
diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
index d93ebba..807bd68 100644
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -2223,6 +2223,15 @@ define Device/zbtlink_zbt-wd323
 endef
 TARGET_DEVICES += zbtlink_zbt-wd323
 
+define Device/zsun_sd100
+  SOC := ar9331
+  DEVICE_VENDOR := Zsun
+  DEVICE_MODEL := SD100
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-storage
+  IMAGE_SIZE := 12032k
+endef
+TARGET_DEVICES += zsun_sd100
+
 define Device/zyxel_nbg6616
   SOC := qca9557
   DEVICE_VENDOR := ZyXEL
